src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,244 @@
+        } else {
+            addLast(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+        if (index == size && index != 0) {
+            addLast(value);
+        } else if (index == 0) {
+            addFirst(value);
+        } else if (size / 2 >= index) {
+            addFromHead(value, index);
+        } else {
+            addFromTail(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexGet(index);
+        return size / 2 >= index ? getFromHead(index) : getFromTail(index);
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexGet(index);
+        return size / 2 >= index ? setFromHead(value, index) : setFromTail(value, index);
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> current = findNode(index, null);
+        removeNode(current);
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = findNodeByValue(object);
+        if (Objects.nonNull(current)) {
+            removeNode(current);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void removeNode(Node<T> current) {
+        if (current == head) {
+            head = current.next;
+        } else {
+            current.prev.next = current.next;
+        }
+        if (current == tail) {
+            tail = current.prev;
+        } else {
+            current.next.prev = current.prev;
+        }
+        size--;
+    }
+
+    private Node<T> findNodeByValue(T value) {
+        Node<T> current = head;
+        while (current != null) {
+            if (Objects.equals(current.item, value)) {
+                return current;
+            }
+            current = current.next;
+        }
+        return null;
+    }
+
+    private Node<T> findNode(int index, T value) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(INVALID_INDEX_EXCEPTION + index);
+        }
+
+        Node<T> current = head;
+        if (value != null) {
+            while (current != null) {
+                if (Objects.equals(current.item, value)) {
+                    return current;
+                }
+                current = current.next;
+            }
+        } else {
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            return current;
+        }
+
+        throw new NoSuchElementException(OBJECT_NOT_FOUND_EXCEPTION + value);
+    }
```

Comment:
it is better to separate this method into 2 different methods: 
1. findNodeByIndex
2. findNodeByValue

that will be much better, because now your method too complicated